Add spacechains-permissionless-blockchains-for-bitcoin review to misc (#676)
bitcointranscripts/bitcointranscripts: misc/spacechains-permissionless-blockchains-for-bitcoin.md

---
title: 'Spacechains – Permissionless Blockchains for Bitcoin'
transcript_by: 'muchai254 via review.btctranscripts.com'
media: 'https://www.youtube.com/watch?v=N2ow4Q34Jeg'
date: '2020-12-13'
tags:
  - 'sidechains'
  - 'covenants'
  - 'sighash-anyprevout'
  - 'op-checktemplateverify'
  - 'eltoo'
  - 'rbf'
speakers:
  - 'Ruben Somsen'
categories:
  - 'Contract Protocols'
  - 'Soft Forks'
  - 'Mining'
source_file: 'https://www.youtube.com/watch?v=N2ow4Q34Jeg'

> summary: "Ruben Somsen presents Spacechains, a design for permissionless, trustless secondary blockchains that extend the Bitcoin ecosystem without creating competing altcoins. The design combines two primitives: Blind Merged Mining (BMM) and the Perpetual One-Way Peg (POWP).\nBMM lets anyone produce a spacechain block by winning a fee-bidding auction for a single, unique slot in the Bitcoin chain, requiring only one Bitcoin transaction per block. It is built from a pre-generated sequence of covenant transactions linked by a one-block relative timelock, using SIGHASH_ANYPREVOUT (or alternatively OP_CHECKTEMPLATEVERIFY) to avoid the circular-reference problem of self-referencing signatures, with users bidding via SIGHASH_ANYONECANPAY + SIGHASH_SINGLE and RBF. This fully decouples spacechain mining from Bitcoin miners, who simply select the highest-fee bid, preserving censorship resistance and inheriting Bitcoin's difficulty and reorg properties.\nPOWP lets users burn BTC for an equivalent amount of \"spacecoins,\" a one-way peg that preserves the 21M cap and gives the token a single, non-speculative purpose: paying spacechain fees."

Reviewer: ```suggestion
summary: "Ruben Somsen presents Spacechains, a design for permissionless, trustless secondary blockchains that extend the Bitcoin ecosystem without creating competing altcoins. The design combines two primitives: Blind Merged Mining (BMM) and the Perpetual One-Way Peg (P1WP).\nBMM lets anyone produce a spacechain block by winning a fee-bidding auction for a single, unique slot in the Bitcoin chain, requiring only one Bitcoin transaction per block. It is built from a pre-generated sequence of covenant transactions linked by a one-block relative timelock, using SIGHASH_ANYPREVOUT (or alternatively OP_CHECKTEMPLATEVERIFY) to avoid the circular-reference problem of self-referencing signatures, with users bidding via SIGHASH_ANYONECANPAY + SIGHASH_SINGLE and RBF. This fully decouples spacechain mining from Bitcoin miners, who simply select the highest-fee bid, preserving censorship resistance and inheriting Bitcoin's difficulty and reorg properties.\nP1WP lets users burn BTC for an equivalent amount of \"spacecoins,\" a one-way peg that preserves the 21M cap and gives the token a single, non-speculative purpose: paying spacechain fees."
```

Reviewer: <img width="646" height="949" alt="Image" src="https://github.com/user-attachments/assets/a651c910-58dc-4dbd-8e12-2421f7e4a245" />

It's not a trivial problem to really ensure that the blockchain moves forward, but it is definitely a solvable problem.
So at the very least, it's not a showstopper.
Okay, so the burning Bitcoin part, which I call the perpetual one-way peg.
One-way peg meaning as opposed to two-way peg, you can move your Bitcoins to this other chain, but one-way so you can never move back, which is why it's a burn, right?
You destroy your coins and you get some spacechain tokens.
And it's perpetual, meaning that you can always do this.

> ## Perpetual One-Way Peg (P1WP)

Reviewer: Rigorous verification! Good job.

Author: Thank you!

And it's actually intended for kind of improving the Lightning Network through something called `eltoo`.
And It actually turns out that for this specific covenant that I'm trying to create, it is slightly more flexible than `op_checktemplateverify`.
So `op_checktemplateverify` and `sighash_anyprevout` turns out they can do roughly the same thing.
`op_checktemplateverify` is usually the cleaner way of doing it.
But in this case, I think there's an argument to be made for `sighash_anyprevout` because basically the transactions become a little smaller and things are a little bit more flexible and we'll see that in a minute.
So I'm going to go and take you through the `sighash_anyprevout` method of creating this countenant.

## Pubkey Spend Transaction (nothing special)

Ruben Somsen: 00:35:49

So I'm starting really simple here.
And I guess I should note this `sighash_anyprevout` was not intended to do the covenant, but it just turned out that it happens to enable this.
So it's actually kind of cool.
And this is something that Anthony Towns taught me.
So yeah, I'll take you through it, kind of how it works.

> So imagine just a regular output, a very simple one where there's a OP key and a CHECKSIG

Reviewer: ```suggestion
So imagine just a regular output, a very simple one where there's a pubkey and a CHECKSIG
```

Reviewer: <img width="1431" height="1330" alt="Image" src="https://github.com/user-attachments/assets/a2af4437-a672-4b6c-9e55-333cd25ef41e" />

## Potential Future

Ruben Somsen: 00:49:03

Just to give you a potential future idea of where this could potentially be headed, and obviously I don't know, but this is what I imagine would be good for Bitcoin.
You have the Bitcoin blockchain, roughly two megabytes.

> I know it's like 4,000 weight units or something like that, but roughly two megabytes worth of data.

Reviewer: ```suggestion
I know it's like 4,000,000 weight units or something like that, but roughly two megabytes worth of data.
```

Reviewer: It's more common to say `4 million weight units`, though 4,000 **kilo** weight units is correct as well.

https://www.spark.money/glossary/block-size-limit
https://www.spark.money/tools/bitcoin-transaction-size-reference
> Blocks are limited to 4,000,000 weight units. Because typical transactions contain roughly 60% witness data by byte count, the effective block size with SegWit transactions is approximately 1.7 to 2.0 MB, compared to the pre-SegWit 1 MB limit.

<img width="1585" height="1426" alt="image" src="https://github.com/user-attachments/assets/251dc343-f7b2-4cc3-962d-b730ceb435ff" />

Reviewer: @0tuedon The speaker either stated the wrong number or mixed up the metric prefixes. Even though it's not what the speaker said, I believe we should correct the technical fact. What do you think?

Reviewer 2: Yh  but we can leave a not with something like *(what the speaker said originally)*

Reviewer 2: just fix this and we are good to merge @muchai254

Author: Attended to it, @0tuedon

So then you have a spacechain, which let's say utilizes the exact same code as Bitcoin does.
Reason for that being that we want people to feel secure in using that chain.
And then from that chain, you create your other chains.
And this is essentially to minimize the amount of Bitcoin space you use.
And maybe you create a big blockchain like we talked about with 32 megabytes worth of data or something like that.
And what you have here is on the Bitcoin blockchain there is only one transaction per block, right?
And that creates the spacechain.
And then from that spacechain, we have the exact same construction creating the big block spacechain.
So it's very efficient in terms of how much Bitcoin space you're utilizing here.
So then maybe you have a confidential assets spacechain, which maybe make it a little bit bigger because, I don't know, we feel like that's good enough.
In this chain, you can create your own assets.
They can be confidential, so you don't really see the history of it.
Somebody could issue a federated two-way peg on there if they feel so inclined, et cetera.

> So then we maybe have to use the Tether chain that I just described, where you have some sort of DEX thing, allowing people to basically use the Tether to move in and out of fiat without using an exchange.

Reviewer: ```suggestion
So then we maybe have to use the Tether chain that I just described, where you have some sort of DEX thing, allowing people to basically use USDT Tether to move in and out of fiat without using an exchange.
```

Reviewer: Regarding your `00:49:03` comment, it sounds like "USD" or "USDT". However, this isn't critical, as it doesn't really affect the technical information.

https://www.youtube.com/watch?v=N2ow4Q34Jeg&t=3030s

<img width="1801" height="1379" alt="Image" src="https://github.com/user-attachments/assets/55832f37-a972-4c34-8b3d-6a8b74d8e9b8" />

I think it's a really cool project, but it still utilizes the Bitcoin blockchain directly.
And as block space becomes more scarce, I think those kinds of use cases become priced out.
Right?
Where here you can create a, you know, let's say USD Tether blockchain, where you can still have cheap fees because you can even create 10 USD Tether blockchains if you wanted.

> As opposed to doing USD Tether as a color coin on the Bitcoin blockchain, where if you want to move your USD Tether, you have to pay really high fees because the Bitcoin block space is so scarce.

Reviewer: ```suggestion
As opposed to doing USD Tether as a colored coin on the Bitcoin blockchain, where if you want to move your USD Tether, you have to pay really high fees because the Bitcoin block space is so scarce.
```

Reviewer: https://bitcoin.stackexchange.com/questions/5695/what-are-colored-coins

<img width="1907" height="852" alt="Image" src="https://github.com/user-attachments/assets/6aee91ef-5f4e-4a31-a435-e71dd76c931a" />

So that's roughly kind of my thinking.
My feeling is that we have to move towards things like this.
And that's also what motivates me to work on this.
So yeah, that concludes my presentation.
Thank you very much for listening.

> If you're interested in talking more about this or just discussing it with me or whatever, please join me on Telegram at a channel I created called t.me.spacechains.

Reviewer: ```suggestion
If you're interested in talking more about this or just discussing it with me or whatever, please join me on Telegram at a channel I created called t.me/spacechains.
```

Reviewer: <img width="760" height="1004" alt="Image" src="https://github.com/user-attachments/assets/29aec88a-0daf-4a0f-b068-26dfeb26cd5f" />
